src/caffe/layer_factory.cpp

```diff
@@ -160,14 +160,18 @@ REGISTER_LAYER_CREATOR(TanH, GetTanHLayer);
 #ifdef WITH_PYTHON_LAYER
 template <typename Dtype>
 shared_ptr<Layer<Dtype> > GetPythonLayer(const LayerParameter& param) {
+  string module_name = param.python_param().module();
+  string layer_name = param.python_param().layer();
   Py_Initialize();
   try {
-    bp::object module = bp::import(param.python_param().module().c_str());
-    bp::object layer = module.attr(param.python_param().layer().c_str())(param);
+    bp::object globals = bp::import("__main__").attr("__dict__");
+    bp::exec(("import " + module_name).c_str(), globals, globals);
+    bp::object layer_class = bp::eval((module_name + "." + layer_name).c_str(),
```

Comment:
Agreed with @seanbell that if we're going to `eval`/`exec`it's worth doing a check like the one he suggests.  Is it necessary to switch away from `bp::import` though?  I'm not sure why it's needed, is it to load `__dict__` first? Could `param_str` just be a string, and then Python layer writers could `eval ` it as a dict on their own if they want it to take a dict?